src/components/ListV2/ListRow.tsx

```diff
@@ -0,0 +1,306 @@
+  }
+
+  & ${StyledIcon} {
+    transition: transform 250ms ease-in-out;
+  }
+
+  &[aria-expanded='true'] ${StyledIcon} {
+    transform: rotate(-180deg);
+  }
+`
+
+type ListRowProps = {
+  children: ReactNode
+  className?: string
+  isHoverable?: boolean
+  isDisabled?: boolean
+  isHighlighted?: boolean
+  isExpandable?: boolean
+  isExpanded?: boolean
+  id: string
+  checkboxRender?: ReactNode
+  checkboxDisabled?: boolean
+  checkboxTooltip?: string
+  tooltip?: string
+  variant?: typeof LIST_ROW_VARIANTS[number]
+  hideArrow?: boolean
+}
+
+export const ListRow = ({
+  children,
+  className,
+  isDisabled,
+  isHoverable = true,
+  isHighlighted,
+  isExpanded: forceExpand,
+  isExpandable,
+  id,
+  checkboxRender,
+  variant,
+  hideArrow = false,
+  tooltip,
+  checkboxTooltip,
+  checkboxDisabled,
+}: ListRowProps) => {
+  const {
+    autoClose,
+    template,
+    selectable,
+    selectedIds,
+    setSelectedIds,
+    expandedIds,
+    setExpandedIds,
+    disabledRowsRef,
+  } = useListContext()
+  const isSelected = id ? selectedIds.includes(id) : false
+  const isExpanded = forceExpand || (id ? expandedIds.includes(id) : false)
+
+  useEffect(() => {
+    if (forceExpand && id && !expandedIds.includes(id)) {
+      setExpandedIds(current => Array.from(new Set([...current, id])))
+    }
+  }, [expandedIds, forceExpand, id, setExpandedIds])
+
+  const handleExpand: MouseEventHandler<HTMLTableRowElement> = () => {
+    setExpandedIds(current => {
+      const indexOfItem = current.indexOf(id)
+      if (indexOfItem > -1) {
+        current.splice(indexOfItem, 1)
+
+        return [...current]
+      }
+
+      return autoClose ? [id] : Array.from(new Set([...current, id]))
+    })
+  }
+
+  const handleCheck: ChangeEventHandler<HTMLInputElement> = event => {
+    if (event.target.checked) {
+      setSelectedIds(current =>
+        Array.from(new Set([...current, event.target.value])),
+      )
+    } else {
+      setSelectedIds(current => {
+        const indexOfItem = current.indexOf(event.target.value)
+        current.splice(indexOfItem, 1)
+
+        return [...current]
+      })
+    }
+  }
+
+  const computedTemplate = isExpandable ? `${template} 25px` : template
+
+  useEffect(() => {
+    if (
+      id &&
+      (isDisabled || checkboxDisabled) &&
+      !disabledRowsRef.current.includes(id)
+    ) {
+      disabledRowsRef.current.push(id)
+    }
+    if (
+      id &&
+      !isDisabled &&
+      !checkboxDisabled &&
+      disabledRowsRef.current.includes(id)
+    ) {
+      disabledRowsRef.current = disabledRowsRef.current.filter(
+        disabledId => disabledId !== id,
+      )
+    }
+  }, [isDisabled, id, disabledRowsRef, checkboxDisabled])
+
+  useEffect(() => {
+    if ((isDisabled || checkboxDisabled) && selectedIds.includes(id)) {
+      setSelectedIds(current => {
+        const indexOfItem = current.indexOf(id)
+        current.splice(indexOfItem, 1)
+
+        return [...current]
```

Comment:
Why mutate current array then copy in a new one? If the purpose was to preserve orginal array, it will maybe  be better to do the oppsite ` return [...current].splice(indexOfItem, 1)` or just returning mutated array.

Reply:
splice return an array of removed elements not the remaining ones